.github/workflows/build_and_test.yml

```diff
@@ -31,6 +31,7 @@ jobs:
           apk update && apk add cppcheck python3-dev
           python3 -m pip install cmake-format clang-format==18.1.6
 
+      # TODO: investigate how to run pre-commit with `venv`
       - uses: pre-commit/action@2c7b3805fd2a0fd8c1884dcaebf91fc102a13ecd # v3.0.1
```

Comment:
Did you try 
```suggestion
      - name: Activate virtualenv
        run: |
          ${{ matrix.python }} -m venv ./venv
          . .venv/bin/activate
          echo PATH=$PATH >> $GITHUB_ENV

      - uses: pre-commit/action@2c7b3805fd2a0fd8c1884dcaebf91fc102a13ecd # v3.0.1
```

Reply:
@tigercosmos it shoul be "source ./venv/bin/activate"

Reply:
`source` is only for bash.  However, I cannot control to let pre-commit action to use bash. I will just convert the change, and I don't see any issue that pre-commit doesn't run in the venv.